### PR TITLE
Allow itermonomials to respect non-commutative variables

### PR DESCRIPTION
**Pull request by dps7ud**
Example:
\>\>\> from sympy import itermonomials, symbols
\>\>\> itermonomials(symbols('a b', commutative=False), 2)
{1, a, b, a\*b, b\*a, a\*\*2, b\*\*2}

Previous implementation would yield:
{1, a, b, a\*b, a\*\*2, b\*\*2}

Efficiency Tests:

from timeit import timeit

timeit("itermonomials( symbols('a b'), 2)",
    setup="from sympy import itermonomials, symbols",
    number=10000)

timeit("itermonomials( symbols('a b'), 7)",
    setup="from sympy import itermonomials, symbols",
    number=10000)

timeit("itermonomials( symbols('a b c'), 2)",
    setup="from sympy import itermonomials, symbols",
    number=10000)

timeit("itermonomials( symbols('a b c'), 4)",
    setup="from sympy import itermonomials, symbols",
    number=10000)

old:
    0.5966764690001582
    4.53667760999997
    1.209770285000559
    4.8836658829986845

new:
    0.22628378899935342
    0.596607915000277
    0.30169753399968613
    0.7843536450000101

### CODE REVIEW

**Comment by gxyd** [comment]
This is related to `polys` module, @jksuom would you be interested in reviewing it?

File: sympy/polys/monomials.py [hunk 1]
```diff
@@ -44,18 +48,20 @@ def itermonomials(variables, degree):
...
+        >>> itermonomials([a, b, x], 2)
+        {1, a, a**2, b, b**2, x, x**2, a*b, b*a, x*a, x*b}
 
-        for i in range(1, degree + 1):
-            monoms |= set([ x**i * m for m in itermonomials(tail, degree - i) ])
 
-        return monoms
+    """
+    if not variables or degree < 1:
+        return {S(1)}
```
Comment (jksuom) [inline sympy/polys/monomials.py]: Maybe the empty set should be returned when `degree < 0`.

Comment (dps7ud) [reply sympy/polys/monomials.py]: That sounds reasonable. Okay to change behavior from previous implementation?

Comment (jksuom) [reply sympy/polys/monomials.py]: I think it is ok in this case. There are no monomials of degree < 0.

**Comment by jksuom** [comment]
Thanks, this looks good. The iterative implementation (in C) of `itertools` is clearly more efficient than the original.

**Pull request merged**